src/bin/miri.rs

```diff
@@ -63,6 +85,34 @@ impl rustc_driver::Callbacks for MiriCompilerCalls {
     }
 }
 
+struct OverrideExportedSymbols {
+    target_crate: bool,
+}
+
+impl rustc_driver::Callbacks for OverrideExportedSymbols {
+    fn config(&mut self, config: &mut Config) {
+        if config.opts.prints.is_empty() && self.target_crate {
+            config.override_queries = Some(|_, local_providers, _| {
+                // `exported_symbols()` provided by rustc always returns empty result if
+                // `tcx.sess.opts.output_types.should_codegen()` is false.
+                local_providers.exported_symbols = |tcx, cnum| {
+                    assert_eq!(cnum, LOCAL_CRATE);
+                    tcx.arena.alloc_from_iter(
+                        tcx.reachable_set(LOCAL_CRATE).iter().filter_map(|&local_def_id| {
+                            tcx.codegen_fn_attrs(local_def_id)
+                                .contains_extern_indicator()
+                                .then_some((
+                                    ExportedSymbol::NonGeneric(local_def_id.to_def_id()),
+                                    SymbolExportLevel::C,
+                                ))
+                        })
```

Comment:
I assume this is copied from somewhere in rustc? Could you add a comment pointing to the original source?